compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1766,24 +1770,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     ) -> (
         &'b mut [&'c mut Candidate<'pat, 'tcx>],
         FxIndexMap<TestBranch<'tcx>, Vec<&'b mut Candidate<'pat, 'tcx>>>,
+        FxIndexMap<TestBranch<'tcx>, Span>,
     ) {
         // For each of the possible outcomes, collect vector of candidates that apply if the test
         // has that particular outcome.
         let mut target_candidates: FxIndexMap<_, Vec<&mut Candidate<'_, '_>>> = Default::default();
+        let mut matched_spans = FxIndexMap::default();
 
         let total_candidate_count = candidates.len();
 
         // Sort the candidates into the appropriate vector in `target_candidates`. Note that at some
         // point we may encounter a candidate where the test is not relevant; at that point, we stop
         // sorting.
         while let Some(candidate) = candidates.first_mut() {
-            let Some(branch) =
+            let Some((branch, matched_span)) =
                 self.sort_candidate(match_place, test, candidate, &target_candidates)
             else {
                 break;
             };
             let (candidate, rest) = candidates.split_first_mut().unwrap();
-            target_candidates.entry(branch).or_insert_with(Vec::new).push(candidate);
+            target_candidates
+                .entry(branch)
+                .or_insert_with(|| {
+                    // By here the candidate is the first candidate for this branch. The matched span must be from
+                    // one of its match pairs if existed.
```

Comment:
```suggestion
                    // The candidate is the first candidate for this branch. If the corresponding match_pair
                    // was fully matched, we record its span here for coverage tracking.
```

Reply:
From looking at the wikipedia article for MCDC, it sounds like we should be tracking every possible outcome of a decision, no? In that case, shouldn't we record all possible branches, even if the match pair was not fully matched?

Reply:
Well, I might be confused about what "decision" means. The most logical would be to track each branch, including sub-branches due to or-patterns. If that's the case, then what this is doing is incorrect. Instead, I'd suggest considering the whole "match" as one condition, with one outcome per subcandidate. You'd collect the information you need in `lower_match_tree`, where `block` is the start block and the block for each outcome is the `subcandidate.pre_binding_block`. Is that what would make sense?

Reply:
> From looking at the wikipedia article for MCDC, it sounds like we should be tracking every possible outcome of a decision, no? In that case, shouldn't we record all possible branches, even if the match pair was not fully matched?

Yes. I was not clear about how to process the branches not fully matched until understand what `full_matched` means from your reviews. Now I modified it a bit so we could handle not fully matched branches and generate appropriate results for more patterns.

> Instead, I'd suggest considering the whole "match" as one condition, with one outcome per subcandidate.
Let's clarify it more. Consider such code 
```rust
// val: Pat
match val {
    Pat::A | Pat::B => { /* Arm 1 */ }
    Pat::C  => { /* Arm 2 */ }
    _ => { /* Arm 3 */ }
}
```
Here `Pat::A | Pat::B`, `Pat::C` are decisions, their outcome mean "whether these pattern is matched or not". In another word, outcome of decision `Pat::A | Pat:B` is `matches!(val, Pat::A | Pat::B)` is true or false and the conditions are sub patterns `Pat::A`, `Pat::B`. By setting `val` to `Pat::A`, we show the outcome is true. By setting `val` to `Pat::C`, we show the outcome is false. Thus we say condition `Pat::A` can effect the outcome of decision because the outcome varies as evaluated value of `val == Pat::A` changes.

Note that the last pattern `_` is not a decision because it never fails to match `val`.

Taking the whole `match` as a decision (do I misunderstand?) might be a bit avant-garde. To some extent I love this view but this means a decision might have more than two outcome, which is not compatible with traditional definition only caring boolean value and is not supported by llvm implementation.

This pr just generate branch coverage report here but I should try to investigate if we could find a way to implement mcdc defined above.

Reply:
You need to handle decisions with multiple outcomes anyway, since individual tests (e.g. enum variant test) can have many branches. How do you deal with that in the current state of the code?

This example is good: could you spell out *precisely* what conditions and decisions you expect to see in this example?

Reply:
> This example is good: could you spell out _precisely_ what conditions and decisions you expect to see in this example?

Sure let's introduce what decisions and conditions need for mcdc first:
```rust
struct Decision {
    // Span of the whole decision, which should contain all spans of its conditions
    span: Span,
    // Total number of conditions in this decision
    conditions_num: usize,
}

struct Condition {
    // Span of this condition, should be contained by its decision
    span: Span,
    // Id of this condition, starting from 1, should be unique in the decision. Can be assigned arbitrarily except that the first evaluated condition should be assigned to 1
    condition_id: usize,
    // Id of the next condition to be evaluated if this condition is true/matched. 0 if the evaluation ends here.
    true_next_id: usize,
    // Id of the next condition to be evaluated if this condition is false/unmatched. 0 if the evaluation ends here.
    false_next_id: usize,
}
```

For code like 
```rust
// val: (Pat, Pat)
match val {
    (Pat::A | Pat::B, Pat::C) => { /* Arm 1 */ } // Suppose this is line 3
    (Pat::D, Pat::C)  => { /* Arm 2 */ } // Suppose this is line 4
    _ => { /* Arm 3 */ }
}
```
For decision `(Pat::A | Pat::B, Pat::C)` we expect (use string to represent span here) 
```rust
let decision = Decision {
    span: "(Pat::A | Pat::B, Pat::C)", 
    conditions_num: 3,
}
let condition_1 = Condition {
    span: "Pat::C", // This is "Pat::C" in line 3
    condition_id: 1, // This condition is the first to be checked
    true_next_id: 2, // Next we check "Pat::A" if this is true
    false_next_id: 0
}
let condition_2 = Condition {
    span: "Pat::A",
    condition_id: 2,
    true_next_id: 0, // We matched all here
    false_next_id: 3 // Next we check "Pat::B" if this is false
}
let condition_3 = Condition {
    span: "Pat::B",
    condition_id: 3,
    // No matter what this condition outcomes, we reach end
    true_next_id: 0,
    false_next_id: 0
}
```
And for decision `(Pat::D, Pat::C)` we expect 
```rust
let decision = Decision {
    span: "(Pat::D, Pat::C)", 
    conditions_num: 2,
}
let condition_1 = Condition {
    span: "Pat::C", // This is "Pat::C" in line 4
    condition_id: 1,
    true_next_id: 2, // Next we check "Pat::D" if this is true
    false_next_id: 0
}
let condition_2 = Condition {
    span: "Pat::D",
    condition_id: 2,
    true_next_id: 0,
    false_next_id: 0
}
```

Reply:
Thank you, that's really helpful! I didn't have a clear idea of the goal here but I think I do now.

I seem to understand that every branch in the match is a decision, and every match pair in a branch is a condition for that branch?

The current way of doing things misses some conditions then I think. I particular the second `C` is missed in your example, because you only record conditions for the first candidate in each branch of a given test. Why not record all the match pairs when they're fully_matched? This would ensure that every match branch gets one condition per match pair.

Off the top of my head, what I'd try is:
- make `sort_candidate` output `Option<(TestBranch, Option<MatchPair>)>` that returns the match pair of it was fully matched;
- make `sort_candidates` return `Map<Branch, Vec<(Candidate, Option<MatchPair>)>>`, to record every fully_matched match pair alongside the candidate it came from;
- when building `target_blocks`, output one condition for each match pair we get.

Does that make sense?

Reply:
Note also that I'm not sure we're handling spans like you want right now: can you try and example like `Some(true) | None`? I suspect the `true` would have an incorrect decision span.

To solve this, I'd suggest accumulating conditions in a new field of `Candidate`, and process them all at the end of `lower_match_tree`. The trick is that you want to use the span of the top-level candidate, but the conditions might be stored in sub-candidates.

Reply:
> Off the top of my head, what I'd try is:
> 
>     * make `sort_candidate` output `Option<(TestBranch, Option<MatchPair>)>` that returns the match pair of it was fully matched;
> 
>     * make `sort_candidates` return `Map<Branch, Vec<(Candidate, Option<MatchPair>)>>`, to record every fully_matched match pair alongside the candidate it came from;
> 
>     * when building `target_blocks`, output one condition for each match pair we get.
> 

Yes that's exactly I am going to do. I would try to find a way to identify each decision alongside lowering the match tree and group its conditions respectively.

> can you try and example like `Some(true) | None`? I suspect the `true` would have an incorrect decision span.

Decision `Some(true) | None` would have 3 conditons: `val.is_some()`, `val.is_none()`, `<val as Option::Some>.0 == true`. To simplify it we can take it as something like `(val.is_some() && <val as Option::Some>.0 == true) || val.is_none()`, then the decision span is `Some(true) | None` and conditions are
```
let condition_1 = Condition {
    span: "Some(" // Primarily is `Some(true)` but we extract the span of its sub pattern from it.
    condition_id: 1,
    true_next_id: 2,
    false_next_id: 3
    // omit blocks
}
let condition_2 = Condition {
   span: "true"
   condition_id: 2,
   true_next_id: 0,
   false_next_id: 0,
}
let condition_3 = Condition {
   span: "None"
   condition_id: 3,
   true_next_id: 0,
   false_next_id: 0,
}
```
For now the decision span is right because I get it by combining spans of all conditions with `Span::to`. Indeed this method would not work if we record all target spans.

> To solve this, I'd suggest accumulating conditions in a new field of `Candidate`, and process them all at the end of `lower_match_tree`. The trick is that you want to use the span of the top-level candidate, but the conditions might be stored in sub-candidates.

Agree with that, I should have a try in this way. Thank you!